week17/controlledForm/components/form.js

```diff
@@ -0,0 +1,150 @@
+};
+
+const isFormComplete = (formData) => {
+    const fieldsCompleted = _.map(formData, (data) => data.length > 0 );
+    return _.every(fieldsCompleted, (field)=> field);
+};
+
+
+class Form extends Component {
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            formData: {
+                title: "Mr",
+                firstname: "",
+                surname: "",
+                email: "",
+                password1: "",
+                password2: "",
+            },
+            buttonEnabled: false,
+            formSubmitted: false,
+            validationMessages: []
+        };
+    }
+
+    handleSubmit(e) {
+        e.preventDefault();
+        console.table(this.state.formData);
+
+        // fake a 200 response
+        this.setState({
+            formSubmitted: true
+        })
+    }
+
+    checkFormCompletion() {
+        const { formData, validationMessages } = this.state;
+
+        this.setState({
+            buttonEnabled: isFormComplete(formData) && validationMessages.length === 0
+        });
+    }
+
+    validation() {
+        const { password1, password2 } = this.state.formData;
+        const passwordsLengthMsg = isLengthValid(password1, "Password", 6);
+        const passwordAlphanumericMsg = isAlphanumeric(password1, "Password");
+        const passwordsMatchMsg = isMatchValid(password1, password2, "Password");
+        // Add any other validation here
+
+        this.setState({
+            validationMessages: [...passwordsMatchMsg, ...passwordsLengthMsg, ...passwordAlphanumericMsg]
+        }, () => {
+            this.checkFormCompletion();
+        });
+    }
+
+    handleChange(e) {
+        const { value, name } = e.target;
+
+        this.setState({
+            formData: { ...this.state.formData, [name]: _.trimStart(value) }
+            }, () => {
+            this.validation();
+        });
+    }
+
+    titleOptions() {
+        return [{value: "mr", text:"Mr"},
+            {value: "Mrs", text:"Mrs"},
+            {value: "miss", text:"Miss"},
+            {value: "dr", text:"Dr"}]
+    }
+
+    render() {
+        const { title, firstname, surname, email, password1, password2 } = this.state.formData;
+        const { buttonEnabled, formSubmitted, validationMessages } = this.state;
+        
+        return (
+            <div className="container">
+                <form className="form" onSubmit={(e) => { this.handleSubmit(e)}}>
+                    <h1 className="form__header">Would you like to know more?</h1>
+                    <fieldset className="form_fieldset">
+                        <FormSelect
+                            options={ this.titleOptions() }
+                            value={ title }
+                            label="Title"
+                            id="title"
+                            callBack={(e) => { this.handleChange(e)}} />
```

Comment:
This is an anti-pattern and will cause a re-render because you are always returning a new function. rather make `this.handleChange` into an arrow function and just reference it directly.

Same with the others below